frontend/src/components/common/Resource/Resource.tsx

```diff
@@ -894,3 +895,175 @@ export function ConditionsSection(props: { resource: KubeObjectInterface | null
+}
+
+export interface PrintVolumeLinkProps {
+  name?: string;
+  configMap?: {
+    name: string;
+    defaultMode: number;
+  };
+  persistentVolumeClaim?: {
+    claimName: string;
+    readOnly: boolean;
+  };
+}
+
+export function VolumeSection(props: VolumeSectionProps) {
+  const { resource } = props;
+
+  const volumes = resource?.spec?.volumes;
+
+  if (!volumes) {
+    return null;
+  }
+
+  const namespace = resource?.metadata?.namespace;
+
+  const { t } = useTranslation();
+
+  /*
+   * printVolumeLink will print a working link that is set within the router using fields from the resource as params
+   */
+  function printVolumeLink(volumeName: string, volume: PrintVolumeLinkProps) {
+    const { persistentVolumeClaim, configMap } = volume;
+
+    if (persistentVolumeClaim) {
+      return (
+        <Link
+          routeName="persistentVolumeClaim"
+          params={{ namespace: namespace, name: persistentVolumeClaim.claimName }}
+        >
+          {persistentVolumeClaim.claimName}
+        </Link>
+      );
+    } else if (configMap) {
+      return (
+        <Link routeName="configMap" params={{ namespace: namespace, name: configMap.name }}>
+          {configMap.name}
+        </Link>
+      );
+    } else {
+      return volumeName;
+    }
+  }
+
+  function volumeRows(volume: VolumeSectionProps['volume']) {
+    const currentVolume = volume;
+    if (!currentVolume) {
+      return [];
+    }
+
+    let volumeName = currentVolume.name;
+    let volumeType: string = '';
+
+    const volumeKeys = Object.keys(currentVolume);
+
+    for (const key of volumeKeys) {
+      if (key === 'name') {
+        volumeName = currentVolume[key];
+      } else {
+        volumeType = key;
+      }
+    }
+
+    function volumeDetailGrab(vol: VolumeSectionProps['volume']) {
+      // array for items that are printable
+      const directPrint = [];
+
+      // array for items that are not printable and need to be printed to yaml
+      const yamlPrint = [];
+
+      // loop over volumeKeys and check if the value is a string, number, or bool
+      for (const key in vol) {
+        if (!(vol[key] === '')) {
+          if (
+            typeof vol[key] === 'string' ||
+            typeof vol[key] === 'number' ||
+            typeof vol[key] === 'boolean'
+          ) {
+            directPrint.push({
+              volKey: key,
+              volValue: vol[key],
+            });
+          } else {
+            yamlPrint.push({
+              volKey: key,
+              volValue: vol[key],
+            });
+          }
+        }
+      }
+
+      const volumePrints = {
+        directPrint,
+        yamlPrint,
+      };
+
+      return volumePrints;
+    }
+
+    const volumeDetails: any = volumeDetailGrab(currentVolume[volumeType]);
+
+    function yamlPrinter(yamlData: any) {
+      return (
+        <Typography component="pre" variant="body2">
+          {YAML.stringify(yamlData)}
+        </Typography>
+      );
+    }
```

Comment:
I think it's a bit hard to read the function having all these sub-functions scattered around the code.